lib/phoenix/pubsub.ex

```diff
@@ -1,170 +1,140 @@
+
+  PubSub functions send the following messages:
+
+    * `subscribe` -
+       sends:        `{:subscribe, pid, topic, link}`
+       respond with: `:ok | {:error, reason} {:perform, {m, f a}}`
+
+    * `unsubscribe` -
+       sends:        `{:unsubscribe, pid, topic}`
+       respond with: `:ok | {:error, reason} {:perform, {m, f a}}`
+
+    * `subscribers` -
+       respond with: `:ok | {:error, reason} {:perform, {m, f a}}`
+
+    * `broadcast` -
+       sends          `{:broadcast, :none, topic, message}`
+       respond with: `:ok | {:error, reason} {:perform, {m, f a}}`
+
+  Additionally, adapters must implement `start_link/2` with the following format:
+
+      def start_link(pubsub_server_name_to_locally_register, options)
+
+
+  ### Offloading work to clients via MFA response
+  The `Phoenix.PubSub` API allows any of its functions to handle a
+  response from the adapter matching `{:perform, {m, f a}}`. The PubSub
+  client will recursively invoke all MFA responses until a result is
+  return. This is useful for offloading work to clients without blocking
+  in your PubSub adapter. See `Phoenix.PubSub.PG2` for an example usage.
+
+
   ## Example
 
-      iex> PubSub.subscribe(self, "user:123")
+      iex> PubSub.subscribe(MyApp.PubSub, self, "user:123")
       :ok
       iex> Process.info(self)[:messages]
       []
-      iex> PubSub.subscribers("user:123")
+      iex> PubSub.subscribers(MyApp.PubSub, "user:123")
       [#PID<0.169.0>]
-      iex> PubSub.broadcast "user:123", {:user_update, %{id: 123, name: "Shane"}}
+      iex> PubSub.broadcast MyApp.PubSub, "user:123", {:user_update, %{id: 123, name: "Shane"}}
       :ok
       iex> Process.info(self)[:messages]
       {:user_update, %{id: 123, name: "Shane"}}
 
   """
 
-  @server Phoenix.PubSub.Server
-
-  @pg_prefix :phx
-
-  @doc """
-  Creates a topic for pubsub broadcast to subscribers
-
-    * topic_name - The String name of the topic
-
-  ## Examples
-
-      iex> PubSub.create("mytopic")
-      :ok
-
-  """
-  def create(topic_name) do
-    :ok = call {:create, group(topic_name)}
+  defmodule BroadcastError do
+    defexception [:message]
+    def exception(msg) do
+      %BroadcastError{message: "Broadcast failed with #{inspect msg}"}
+    end
   end
 
+
   @doc """
-  Checks if a given topic is registered as a process group
+  Subscribes the pid to the PubSub adapter's topic
+
+    * `server` - The Pid registered name of the server
+    * `pid` - The subscriber pid to receive pubsub messages
+    * `topic` - The topic to subscribe to, ie: `"users:123"`
+    * `opts` - The optional list of options. Supported options
+               only include `:link` to link the subscriber to
+               the pubsub adapter
   """
-  def exists?(topic_name) do
-    call {:exists?, group(topic_name)}
-  end
+  def subscribe(server, pid, topic, opts \\ [])
+  def subscribe(server, pid, topic, link: link),
+    do: call(server, {:subscribe, pid, topic, link && true})
```

Comment:
I think we should pass the options to the server. We won't need to change the API in case we add other options in the future and we won't need to have the link casting logic in multiple places.